src/libraries/VanityAddressLib.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+/// @title VanityAddressLib
+/// @notice A library to score addresses based on their vanity
+library VanityAddressLib {
+    /// @notice Compares two addresses and returns true if the first address has a better vanity score
+    /// @param first The first address to compare
+    /// @param second The second address to compare
+    /// @return better True if the first address has a better vanity score
+    function betterThan(address first, address second) internal pure returns (bool better) {
+        return score(first) > score(second);
+    }
+
+    /// @notice Scores an address based on its vanity
+    /// @param addr The address to score
+    /// @return calculatedScore The vanity score of the address
+    function score(address addr) internal pure returns (uint256 calculatedScore) {
+        //  Scoring rules:
+        //     Requirement: The first nonzero nibble must be 4
+        //     10 points for every leading 0 nibble
+        //     40 points if the first 4 is followed by 3 more 4s
+        //     20 points if the first nibble after the 4 4s is NOT a 4
+        //     20 points if the last 4 nibbles are 4s
+        //     1 point for every 4
```

Comment:
Confirming a design decision, but from the PR description:
> leading 0's are weighted most heavily, followed by other 0's and 4's

While the most optimal address has many leading 0s (`0x000000000000000000000000000000000000004444` with 360 points), `0x4444000000000000000000000000000000000000` gets a score of 64 whereas `0x0004000000000000000000000000000000000000` would only get a score of 31 pts.

The implemented scoring system optimizes for:
1. Leading 0s only if there are more than 6 of them. (> 60pts)
2. a group of 4s after the first zero (or at the beginning if there is no leading 0) (60 pts)
3. fewer than 6 leading 0s (10-60 pts)
4. 4 4s at the end (20 pts)
5. Other 4s that don't qualify for 2 or 4
It does not give any weight to "other 0's" after the leading ones.

Reply:
yea that sounds right. do you have any suggestions here?